File: scripts/global/auth-profile-enforcer.js
#!/usr/bin/env node
// #2910 — Authorization profile runtime enforcer (G-05, OWASP ASI02)
// Reads $MEGINGJORD_AUTH_PROFILE and returns allow/deny for a requested capability.
// Used by pretool_guard (via Python companion) and directly in JS tests.
'use strict';

const path = require('path');
const { parseActiveProfile, readSchema } = require('./authorization-profile');

const SCHEMA_PATH = path.join(__dirname, '..', '..', 'config', 'authorization-profiles.json');

// Command patterns mapped to required capabilities
const CAPABILITY_PATTERNS = {
  install: [
    /\bnpm\s+(?:install|i|ci)\b/,
    /\bpip3?\s+install\b/,
    /\bapt(?:-get)?\s+install\b/,
    /\bbrew\s+install\b/,
    /\byarn\s+add\b/,
    /\bpnpm\s+(?:install|add)\b/,
    /\bcargo\s+install\b/,
    /\bgo\s+install\b/,
  ],
  privileged: [
    /(?:^|[\s;|&])\bsudo\b/,
    /(?:^|[\s;|&])\bsu\s+-/,
    /\bchmod\b.*\b[0-7]{3,4}\b/,
    /\bchown\b/,
  ],
  execute_remote: [
    /(?:^|[\s;|&])\bssh\b\s+\S/,
    /(?:^|[\s;|&])\bscp\b\s+\S/,
    /\brsync\b[^|&\n]*\w@[\w.]+:/,
    /(?:^|[\s;|&])\bansible(?:-playbook)?\b/,
    /(?:^|[\s;|&])\bfabric\b/,
  ],
};

/**
 * Infer which capabilities a shell command requires.
 * Returns an array of capability names (may be empty).
 */
function inferCapabilities(command) {
  const needed = [];
  for (const [cap, patterns] of Object.entries(CAPABILITY_PATTERNS)) {
    if (patterns.some(re => re.test(command))) {
      needed.push(cap);
    }
  }
  return needed;
}

/**
 * Check whether a named capability is permitted under the active profile.
 * @param {string} capability - one of install|upgrade|privileged|execute_local|execute_remote
 * @param {object} [opts] - override argv/env/schema for testing
 * @returns {{ allowed: boolean, profile: string, capability: string, reason: string }}
 */
function checkCapability(capability, opts = {}) {
  try {
    const schema = opts.schema || readSchema(SCHEMA_PATH);
    const active = parseActiveProfile({ schema, env: opts.env || process.env, argv: opts.argv || [] });
    const allowed = active.capabilities[capability] !== false;
    const reason = allowed
      ? `Profile '${active.profile}' permits '${capability}'`
      : `Profile '${active.profile}' blocks '${capability}' (G-05, ASI02)`;
    return { allowed, profile: active.profile, capability, reason };
  } catch (e) {
    // Fail-open: enforcer errors must never brick the gate
    return { allowed: true, profile: 'unknown', capability, reason: `enforcer-error: ${e.message}` };
  }
}

/**
 * Check a shell command string against the active authorization profile.
 * Returns the first deny decision found, or null if all capabilities are permitted.
 * @param {string} command - raw command string
 * @param {object} [opts] - override opts for testing
 * @returns {{ allowed: false, profile, capability, reason } | null}
 */
function checkCommand(command, opts = {}) {
  const needed = inferCapabilities(command);
  for (const cap of needed) {
    const result = checkCapability(cap, opts);
    if (!result.allowed) return result;
  }
  return null;
}

function main() {
  const args = process.argv.slice(2);
  const capArg = args.find(a => a.startsWith('--capability='));
  const cmdArg = args.find(a => a.startsWith('--command='));
  try {
    let result;
    if (capArg) {
      result = checkCapability(capArg.split('=')[1]);
    } else if (cmdArg) {
      result = checkCommand(cmdArg.split('=').slice(1).join('=')) || {
        allowed: true, profile: 'unknown', capability: 'none', reason: 'no restricted capability detected',
      };
    } else {
      process.stderr.write('Usage: auth-profile-enforcer.js --capability=<name>|--command=<cmd>\n');
      process.exit(2);
    }
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(result.allowed ? 0 : 1);
  } catch (e) {
    process.stderr.write(`auth-profile-enforcer: ${e.message}\n`);
    process.exit(2);
  }
}

module.exports = { inferCapabilities, checkCapability, checkCommand, CAPABILITY_PATTERNS };
if (require.main === module) main();

File: hooks/scripts/auth_profile_enforcer.py
"""#2910 — Authorization profile runtime enforcement (G-05, OWASP ASI02).

Provides capability checks for pretool_guard.py: reads MEGINGJORD_AUTH_PROFILE
from the environment, loads the capability matrix from config, and returns
allow/deny decisions for privileged operations.

Fail-open design: any load error falls back to 'owner' (full authority) so
the enforcer never bricks the hook on a misconfigured environment.
"""
import json
import os
import re
from pathlib import Path

_CONFIG_PATH = Path(__file__).resolve().parent.parent.parent / "config" / "authorization-profiles.json"
_VALID_PROFILES = ("owner", "guarded", "restricted")

# Regex patterns mapping commands to required capabilities
_INSTALL_RE = re.compile(
    r"\bnpm\s+(?:install|i|ci)\b"
    r"|\bpip3?\s+install\b"
    r"|\bapt(?:-get)?\s+install\b"
    r"|\bbrew\s+install\b"
    r"|\byarn\s+add\b"
    r"|\bpnpm\s+(?:install|add)\b"
    r"|\bcargo\s+install\b"
    r"|\bgo\s+install\b"
)
_PRIVILEGED_RE = re.compile(
    r"(?:^|[\s;|&])\bsudo\b"
    r"|(?:^|[\s;|&])\bsu\s+-"
    r"|\bchown\b"
)
_EXECUTE_REMOTE_RE = re.compile(
    r"(?:^|[\s;|&])\bssh\b\s+\S"
    r"|(?:^|[\s;|&])\bscp\b\s+\S"
    r"|\brsync\b[^|&\n]*\w@[\w.]+:"
    r"|(?:^|[\s;|&])\bansible(?:-playbook)?\b"
)


def _load_profile(env: dict | None = None) -> tuple[str, dict]:
    """Load active profile name and capability dict. Fail-open → owner."""
    env = env if env is not None else os.environ
    profile = (env.get("MEGINGJORD_AUTH_PROFILE") or "owner").lower()
    if profile not in _VALID_PROFILES:
        profile = "owner"
    try:
        with open(_CONFIG_PATH, "r", encoding="utf-8") as f:
            schema = json.load(f)
        caps = schema["profiles"][profile]
    except Exception:
        caps = {k: True for k in ("install", "upgrade", "privileged", "execute_local", "execute_remote")}
    return profile, caps


def check_capability(capability: str, env: dict | None = None) -> tuple[bool, str]:
    """Return (allowed, reason) for a named capability under the active profile.

    Fail-open: if the capability name is not in the matrix, return (True, reason).
    """
    profile, caps = _load_profile(env)
    if capability not in caps:
        return True, f"Capability '{capability}' not in profile matrix — skipped"
    allowed = bool(caps[capability])
    if allowed:
        return True, f"Profile '{profile}' permits '{capability}'"
    return False, (
        f"Profile '{profile}' blocks '{capability}' (#2910, G-05, ASI02). "
        f"Switch to MEGINGJORD_AUTH_PROFILE=owner to re-enable."
    )


def check_command(command: str, env: dict | None = None) -> tuple[bool, str] | None:
    """Check a shell command string against the active authorization profile.

    Returns (False, reason) for the first denied capability, or None if all permitted.
    Fail-open: any detection error returns None (never block on enforcer bug).
    """
    try:
        if _INSTALL_RE.search(command):
            allowed, reason = check_capability("install", env)
            if not allowed:
                return False, reason
        if _PRIVILEGED_RE.search(command):
            allowed, reason = check_capability("privileged", env)
            if not allowed:
                return False, reason
        if _EXECUTE_REMOTE_RE.search(command):
            allowed, reason = check_capability("execute_remote", env)
            if not allowed:
                return False, reason
    except Exception:
        pass  # Fail-open: detection errors never block
    return None
